File: algorithms/0036.valid-sudoku/solution.rb
# @param {Character[][]} board
# @return {Boolean}
def is_valid_sudoku(board)
  rows = Array.new(9) { Set.new }
  cols = Array.new(9) { Set.new }
  boxes = Array.new(3) { Array.new(3) { Set.new } }

  board.each_with_index do |row, row_idx|
    row.each_with_index do |x, col_idx|
      next if x == '.'

      if rows[row_idx].include?(x)
        return false
      else
        rows[row_idx] << x
      end

      if cols[col_idx].include?(x)
        return false
      else
        cols[col_idx] << x
      end

      if boxes[row_idx / 3][col_idx / 3].include?(x)
        return false
      else
        boxes[row_idx / 3][col_idx / 3] << x
      end
    end
  end

  true
end

File: algorithms/0036.valid-sudoku/solution.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows: Vec<HashSet<&char>> = vec![HashSet::with_capacity(9); 9];
        let mut cols: Vec<HashSet<&char>> = vec![HashSet::with_capacity(9); 9];
        let mut boxes: Vec<Vec<HashSet<&char>>> = vec![vec![HashSet::with_capacity(9); 3]; 3];

        for (row_idx, row) in board.iter().enumerate() {
            for (col_idx, x) in row.iter().enumerate() {
                if (x == &'.') {
                    continue;
                }

                if rows[row_idx].contains(x) {
                    return false;
                } else {
                    rows[row_idx].insert(x);
                }

                if cols[col_idx].contains(x) {
                    return false;
                } else {
                    cols[col_idx].insert(x);
                }

                if boxes[row_idx / 3][col_idx / 3].contains(x) {
                    return false;
                } else {
                    boxes[row_idx / 3][col_idx / 3].insert(x);
                }
            }
        }

        true
    }
}
